### PR TITLE
Add extensibility to PreviewOptions

### LINKED ISSUE #25309: Allow external preview links in Preview dropdown to be extended
In the [AMP plugin](https://wordpress.org/plugins/amp/), we introduce a separate Preview button to be able to preview how the page appears in AMP. We had to put in place a bunch of hacks to get this to work in the block editor since the area around the preview button is not extensible. See https://github.com/ampproject/amp-wp/pull/4397. The button looks like this (screenshot courtesy of @simison):

> AMP Preview button

Unfortunately, this broke in Gutenberg 8.8. But this is not surprising since it was implemented with hacks. We actually don't want to have a separate AMP preview button in the first place. Ideally the AMP preview link would be exposed in the relatively new device preview dropdown (https://github.com/ampproject/amp-wp/issues/4368#issuecomment-597943645), as also was suggested in https://github.com/Automattic/wp-calypso/issues/42896:

> Preview dropdown 

However, we weren't able to do so given that this dropdown is not extensible. 

So my request is that the Preview dropdown menu items be extensible. Plugins should be able to add their own preview links in addition to "Preview externally". This will be useful not only for AMP, but it could also be used by other plugins that may provide alternate representations of the content (e.g. FB Instant Articles, Apple News) or even as a way to link to preview the content on a site that is using WordPress headlessly.

This relates to https://github.com/WordPress/gutenberg/issues/21235 which is about extending the non-external preview options.

### PR DESCRIPTION
**Pull request by arcangelini**
## Description

**This is a continuation of the following PR #31984**
The overall goal of this PR is to make the Preview menu extensible using SlotFill similarly to `PinnedItems` and `PluginSidebar`.

Markup 2021-11-01 at 16 19 21

The menu is extended using the `<PluginPreview>` component built into `@wordpress/interface package`. It allows for defining a custom "Preview" menu item. There are then two ways to manipulate the preview.

1. Using `children` will completely replace the `VisualEditor` component. In my example, I used an iframe and grabbed the preview URL to display a full site preview in iPhone resolution.
2. Using `wrapper` component, the `VisualEditor` iframe is passed in as a prop and allows you to control how the preview is displayed. In my example, I added a basic div, but this could be more robust if desired.

In absence of both of these the view will default back to 'Desktop' allowing for only adding an `onClick` event with no change to the display. Along with the addition of the SlotFill, this PR also refactors the `<PreviewOptions>` component to consolidate the device type names. 

## How has this been tested?

I created a test plugin to allow for easy testing, [download here](https://github.com/WordPress/gutenberg/files/7498806/preview-tester.zip) (**updated**). Once you install the plugin you should see three new items added to your previews dropdown.

Likewise, you can use the following code to create your own plugin and test as you like:

#### Updated with wrapper prop
```javascript
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { PluginPreview } from '@wordpress/interface';
import { registerPlugin } from '@wordpress/plugins';
import { external } from '@wordpress/icons';
import { store as editorStore } from '@wordpress/editor';
import { useSelect } from '@wordpress/data';

function PluginPreviewTest() {
	const previewLink = useSelect(
		( select ) => select( editorStore ).getEditedPostPreviewLink(),
		[]
	);

	function wrapper( content ) {
		return (
			<>
				<h1
					style={ {
						textAlign: 'center',
						margin: 25,
						textTransform: 'uppercase',
					} }
				>
					My custom wrapper
				</h1>
				<div
					className="Custom"
					style={ {
						padding: 175,
						background: '#03395C',
						width: '100%',
						height: '100%',
					} }
				>
					{ content }
				</div>
			</>
		);
	}

	return (
		<>
			<PluginPreview
				name="preview-custom-1"
				title={ __( 'VisualEditor Wrapper' ) }
				wrapper={ wrapper }
			></PluginPreview>

			<PluginPreview
				name="preview-custom-2"
				title={ __( 'Overwrite VisualEditor' ) }
				onClick={ ( event ) => console.log( event ) }
			>
				<div style={ { background: 'grey', height: '100%' } }>
					<h1
						style={ {
							textAlign: 'center',
							margin: 25,
							textTransform: 'uppercase',
						} }
					>
						Iframe to the page preview URL
					</h1>
					<iframe
						style={ {
							width: 375,
							height: 812,
							margin: 'auto',
							display: 'block',
						} }
						title="Plugin Preview"
						src={ previewLink }
					></iframe>
				</div>
			</PluginPreview>

			<PluginPreview
				name="custom-action"
				title={ __( 'Custom Action' ) }
				icon={ external }
				onClick={ ( event ) => console.log( event ) }
			/>
		</>
	);
}

registerPlugin( 'preview-tester', {
	render() {
		return (
			<>
				<PluginPreviewTest />
			</>
		);
	},
} );
```

## Screenshots 

Screen Capture on 2021-11-08 at 18-06-49

Markup 2021-11-01 at 16 26 33

## Types of changes
New feature (non-breaking change which adds functionality)

Fixes #25309
Relates to Automattic/wp-calypso#42896
Relates to ampproject/amp-wp#5943

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/components/preview-options/index.js [hunk 1]
```diff
@@ -45,28 +48,24 @@ export default function PreviewOptions( {
...
+						{ coreDeviceTypes.map( ( device ) => (
+							<MenuItem
+								key={ device }
+								className="block-editor-post-preview__button-resize"
+								onClick={ () => setDeviceType( device ) }
+								icon={ deviceType === device && check }
+							>
+								{ device }
+							</MenuItem>
+						) ) }
```
Comment (ivan-ottinger) [inline packages/block-editor/src/components/preview-options/index.js]: I like how you introduced the `coreDeviceTypes` array and then streamlined the way the
`MenuItem` components are being rendered using `map()`.

One thing I think we could consider is to make the output of device labels
translatable.

An idea I have on my mind is to turn the array of strings into the array of objects
and let the label have its own translatable value that isn't used for any logic:

```diff
diff --git a/packages/block-editor/src/components/preview-options/index.js b/packages/block-editor/src/components/preview-options/index.js
index 03b3f3cea9..c3e67ab580 100644
--- a/packages/block-editor/src/components/preview-options/index.js
+++ b/packages/block-editor/src/components/preview-options/index.js
@@ -19,7 +19,11 @@ export default function PreviewOptions( {
 	deviceType,
 	setDeviceType,
 } ) {
-	const coreDeviceTypes = [ 'Desktop', 'Tablet', 'Mobile' ];
+	const coreDeviceTypes = [
+		{ type: 'Desktop', label: __( 'Desktop' ) },
+		{ type: 'Tablet', label: __( 'Tablet' ) },
+		{ type: 'Mobile', label: __( 'Mobile' ) },
+	];
 
 	const isMobile = useViewportMatch( 'small', '<' );
 	if ( isMobile ) return null;
@@ -50,12 +54,12 @@ export default function PreviewOptions( {
 					<MenuGroup>
 						{ coreDeviceTypes.map( ( device ) => (
 							<MenuItem
-								key={ device }
+								key={ device.type }
 								className="block-editor-post-preview__button-resize"
-								onClick={ () => setDeviceType( device ) }
-								icon={ deviceType === device && check }
+								onClick={ () => setDeviceType( device.type ) }
+								icon={ deviceType === device.type && check }
 							>
-								{ device }
+								{ device.label }
 							</MenuItem>
 						) ) }
 					</MenuGroup>
```

**Comment by ivan-ottinger** [comment]
Thank you for adding the plugin download link to the PR. It made it much easier to test your code. 👍🏼 👍🏼

**Comment by arcangelini** [comment]
I rebased this PR. I realize this is still a rough draft, but would love feedback to potentially get this moving forward if possible 😄 

@youknowriad @gziolo

**Comment by tomektomczuk** [comment]
I get this error 

export 'PluginPreview' (imported as 'PluginPreview') was not found in '@wordpress/interface' (possible exports: ActionItem, ComplementaryArea, ComplementaryAreaMoreMenuItem, FullscreenMode, InterfaceSkeleton, MoreMenuDropdown, MoreMenuFeatureToggle, PinnedItems, PreferencesModal, PreferencesModalSection, PreferencesModalTabs, ___unstablePreferencesModalBaseOption, store)

**Comment by UranymusDev** [comment]
> 

Yea me too, did you manage to work this out?

**Comment by arcangelini** [comment]
It has been a little while since I have worked on this. I would be happy to clean it up this week and see what is going on. I will let y'all know when it is ready to test again @UranymusDev and @tomektomczuk

**Comment by phil-sola** [comment]
Is there any update on this please? This is a much-needed feature and I know a lot of people are keen to see this happen.

**Comment by arcangelini** [comment]
This is no longer relevant it seems. Closing.

**Comment by gziolo** [comment]
Implemented with #64644.

**Pull request closed without merging**